File: solutions/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int n=height.length;
        int[] left=new int[n];
        int[] right=new int[n];
        int max=0,max1=0,total=0;
        int h=n-1;
        for(int i=0;i<n;i++)
        {
            if(max<height[i])
                max=height[i];
            left[i]=max;

            if(max1<height[h])
                max1=height[h];
            right[h]=max1;
            h--;
        }
        for(int i=0;i<n;i++)
            total+=Math.min(right[i],left[i])-height[i];
        return total;
    }
}

File: solutions/0042-trapping-rain-water/solution.c
int trap(int* height, int heightSize) {
    int left[heightSize];
    int right[heightSize];
    int max=0,sum=0;
    for(int i=0;i<heightSize;i++)
        left[i]=max=fmax(max,height[i]);
    max=0;
    for(int i=heightSize-1;i>=0;i--)
        right[i]=max=fmax(max,height[i]);
    for(int i=0;i<heightSize;i++)
        sum+=(fmin(left[i],right[i])-height[i]);
    return sum;
}
